### PR TITLE
[AERIE-1848] Add expand/clamp transformers to constraints eDSL

### PR DESCRIPTION
**Pull request by adrienmaillard**
* **Tickets addressed:** AERIE-1848
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description

Sibling PR to #185.
I have reimplemented the transformers in the constraints library.

## Verification

New tests.

## Documentation

Documentation about these operators has been added to the constraints page. 

## Future work

Other filters and transformers.

### CODE REVIEW

**Comment by JoelCourtney** [comment]
I have a concern that "clamp" is suggestive of a different operation, and that what you've implemented here is actually "shrink". Usually "clamp" means ensuring an input lies in a range. If the input is outside of the range, the closest edge of the range is returned. If the input is inside the range, it is returned unchanged. I think it could cause confusion if we call this operation "clamp" instead of "shrink"

**Comment by Twisol** [comment]
> Usually "clamp" means ensuring an input lies in a range. If the input is outside of the range, the closest edge of the range is returned. If the input is inside the range, it is returned unchanged.

For instance, [the `clamp` function](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/clamp.xhtml) provided by OpenGL, which is where I learned the terminology.

**Comment by Twisol** [comment]
It seems like `contractBy` and `expandBy` do the same operation, i.e. offsetting the start and end of a window by some amount. They only appear to differ in the assumed sign of the applied offsets. Would it be desirable and/or possible to keep the AST limited to the one fundamental operation (of applying offsets to endpoints)?

**Comment by adrienmaillard** [comment]
@Twisol we could do that. How would call this operation ? `OffsetEndpointsBy` ?

**Comment by mattdailis** [comment]
> @Twisol we could do that. How would call this operation ? `OffsetEndpointsBy` ?

I'm okay with that name, unless someone suggests a better one :+1:

It occurs to me that this can be used to implement the window "shift" operator as well.

**Comment by JoelCourtney** [comment]
I'll vote for just calling it `windows.shift(leftBoundShift, rightBoundShift)`.

**Comment by Twisol** [comment]
I can't seem to come up with a better name for that particular operation. We're already not far off a full convolution operation between Windows though -- maybe we just implement that, call it "convolveWith", and implement all the other common forms in terms of it once and for all?

My internet is out (so I'm on my phone), but let me try to write some pseudocode for how "convolveWith" works on Windows. (This isn't the fully abstract definition of convolution, and it's been months since I worked out the correspondence, but the abstract definition specializes to this one on functions R -> {0, 1}.)

```python
def convolve(windows1, windows2):
  convolution = Windows()
  for x in windows1:
    for y in windows2:
      convolution.add(Window(x.start + y.start, x.end + y.end))
  return convolution
```

Then we can implement  `w.offsetEndpointsBy(left, right)` as `convolve(w, Windows(Window(left, right)))`.

**Comment by Twisol** [comment]
I just worked out the correspondence with convolution for myself, and I think it doesn't admit shrinking windows. So you can safely ignore me --  I'm fine with Joel's name.

**Comment by adrienmaillard** [comment]
`Shift` would be the name of the constraint tree node and `shift` the name of the method in windows. Are we in agreement ? If yes, I am also going to mirror this in the constraint edsl (and replace both shrink/expand nodes by just a shift node).

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/ShiftBy.java
Comment (patkenneally) [inline constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/ShiftBy.java]: :small_red_triangle_down: Should this be "shiftBy" now that we've changed the method name?

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/ShiftBy.java
Comment (patkenneally) [inline constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/ShiftBy.java]: :small_red_triangle_down: Should these parameters be renamed to `shiftFromStart` and `shiftFromEnd` ?

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/json/ConstraintParsers.java
Comment (patkenneally) [inline constraints/src/main/java/gov/nasa/jpl/aerie/constraints/json/ConstraintParsers.java]: :large_orange_diamond: I think private is not needed here?

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/json/ConstraintParsers.java
Comment (patkenneally) [inline constraints/src/main/java/gov/nasa/jpl/aerie/constraints/json/ConstraintParsers.java]: :large_orange_diamond: I think this method should be named `shiftByF` in keeping with the naming convention.

File: merlin-server/constraints-dsl-compiler/src/libs/constraints-ast.ts
Comment (patkenneally) [inline merlin-server/constraints-dsl-compiler/src/libs/constraints-ast.ts]: :large_orange_diamond: There is a `type Duration = number` on line 482. Should we be using that type here?

File: merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java
Comment (patkenneally) [inline merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java]: :large_orange_diamond: Should this be `testShiftBy` ?

File: merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java
Comment (patkenneally) [inline merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java]: :large_orange_diamond: This is a loooooong line, I suggest we add some breaks/formatting for easier parsing.

File: merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java
Comment (patkenneally) [inline merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java]: 🔶 Same here on formatting.

**Review by patkenneally — CHANGES_REQUESTED** [review]
Nicely done.

Legend:
:small_red_triangle_down: *Issues* to address before merge
:large_orange_diamond: *Requests* that should not block merge, but should at least be discussed
:large_blue_circle: *Recommendations* that can be ignored if desired

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/json/ConstraintParsers.java
Comment (JoelCourtney) [reply constraints/src/main/java/gov/nasa/jpl/aerie/constraints/json/ConstraintParsers.java:157]: I general I prefer to restrict visibility as much as possible, much like `const`/`final`-correctness. If it can be private, I think it should be.

File: merlin-server/constraints-dsl-compiler/src/libs/constraints-ast.ts
Comment (JoelCourtney) [reply merlin-server/constraints-dsl-compiler/src/libs/constraints-ast.ts:76]: Seconded: I added that line, it is currently not used. I only added it because that same line existed in `scheduling-ast.ts` too as a placeholder. When we replace `Duration` with a more appropriate type, we'll have to change this to `Duration` anyway.

File: merlin-server/constraints-dsl-compiler/src/libs/constraints-edsl-fluent-api.ts
Comment (JoelCourtney) [inline merlin-server/constraints-dsl-compiler/src/libs/constraints-edsl-fluent-api.ts]: Same comment as Pat's about `number` vs `Duration`.

File: merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java
Comment (JoelCourtney) [inline merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java:286]: (not blocker) Fun fact, you can omit `Real.Value` in the right-hand-side of all operations in the API :) So `.equal(4)` will work too.

File: merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java
Comment (JoelCourtney) [inline merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java]: This test seems redundant since you've merged them into one operation.

**Review by JoelCourtney — CHANGES_REQUESTED** [review]
Love it! Only one real change request that Pat didn't mention, the rest are minor. (see comments) I'm having some trouble with github's comment notifications, so feel free to message me on slack when you're done and I'll get you a quick approval.

File: constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/ShiftBy.java
Comment (adrienmaillard) [reply constraints/src/main/java/gov/nasa/jpl/aerie/constraints/tree/ShiftBy.java]: I've set them to `fromStart` and `fromEnd`.

File: merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java
Comment (adrienmaillard) [reply merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java]: I've renamed it `testShift`. And following Joel's comment, I've removed the other very similar test.

File: merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java
Comment (adrienmaillard) [reply merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java:286]: Nice feature !

File: merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java
Comment (adrienmaillard) [reply merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/services/ConstraintsDSLCompilationServiceTests.java]: Removed it.

**Comment by adrienmaillard** [comment]
I've integrated all of your changes. I've also updated the documentation on the Constraints page.

File: merlin-server/constraints-dsl-compiler/src/libs/constraints-ast.ts
Comment (adrienmaillard) [reply merlin-server/constraints-dsl-compiler/src/libs/constraints-ast.ts]: It is in the constraints edsl, not the ast. But I have added it to the AST and used it in this node.

**Review by JoelCourtney — APPROVED** [review]

**Review by patkenneally — APPROVED** [review]

**Pull request merged**
